grpc/shared_conn.go

```diff
@@ -71,20 +75,70 @@ type SharedConn struct {
 	// `peerConnMu` synchronizes changes to the underlying `peerConn`. Such that calls consecutive
 	// calls to `GrpcConn` and `PeerConn` will return connections from the same (or newer, but not
 	// prior) "generations".
-	peerConnMu     sync.Mutex
-	peerConn       *webrtc.PeerConnection
-	peerConnReady  <-chan struct{}
-	peerConnClosed <-chan struct{}
+	peerConnMu    sync.Mutex
+	peerConn      *webrtc.PeerConnection
+	peerConnReady <-chan struct{}
 	// peerConnFailed gets closed when a PeerConnection fails to connect. The peerConn pointer is
 	// set to nil before this channel is closed.
 	peerConnFailed chan struct{}
 
 	onTrackCBByTrackNameMu sync.Mutex
 	onTrackCBByTrackName   map[string]OnTrackCB
 
+	// isConnectedToViamServer identifies whether this SharedConn is running inside a viam-server
+	// talking to a module, or a module talking to a viam-server. We use this to determine whether
+	// to use long names or short names for PeerConnection video (or audio) track names.
+	isConnectedToViamServer bool
+
 	logger logging.Logger
 }
 
+// NewSharedConnForModule acts as a constructor for `SharedConn` for modules that are communicating
+// back to their parent viam-server.
+func NewSharedConnForModule(grpcConn rpc.ClientConn, peerConn *webrtc.PeerConnection, logger logging.Logger) *SharedConn {
+	// We must be passed a ready connection.
+	pcReady := make(chan struct{})
+	close(pcReady)
+
+	ret := &SharedConn{
+		peerConn:      peerConn,
+		peerConnReady: pcReady,
+		// We were passed in a ready connection. Only create this for when `Close` is called.
+		peerConnFailed:          make(chan struct{}),
+		onTrackCBByTrackName:    make(map[string]OnTrackCB),
+		isConnectedToViamServer: true,
+		logger:                  logger,
+	}
+	ret.grpcConn.ReplaceConn(grpcConn)
+
+	ret.peerConn.OnTrack(func(trackRemote *webrtc.TrackRemote, rtpReceiver *webrtc.RTPReceiver) {
+		ret.onTrackCBByTrackNameMu.Lock()
+		onTrackCB, ok := ret.onTrackCBByTrackName[trackRemote.StreamID()]
+		ret.onTrackCBByTrackNameMu.Unlock()
+		if !ok {
+			msg := "Callback not found for StreamID: %s, keys(resOnTrackCBs): %#v"
+			ret.logger.Errorf(msg, trackRemote.StreamID(), maps.Keys(ret.onTrackCBByTrackName))
+			return
+		}
+		onTrackCB(trackRemote, rtpReceiver)
+	})
+
+	return ret
+}
+
+// IsConnectedToModule returns whether this shared conn is being used to communicate with a module.
+func (sc *SharedConn) IsConnectedToModule() bool {
```

Comment:
I created two methods for this because I didn't think it was obvious this was indeed binary? And I like to read these descriptions in the affirmative tense